### DISCUSSION TITLE
Kubernetes integration

### DISCUSSION THREAD

**Discussion by skar07** [coollabsio/coolify#85] [Feature Requests]
This is an extended thread about how Coolify could _possibly_ integrate Kubernetes and utilize it to its full extent. All of the content presented here is presented in the view of an enterprise / production-ready environment; where Coolify would be expected to scale across thousands of servers. Since the announcement of v2 of Coolify, this would be the right time to consider these ideas and decide on their viability.

Here are the challenges that this Discussion is proposing to address:

- [ ] Scale: across N amount of servers as a single instance
- [ ] Features: Cron jobs (Background jobs), Building with Cloud Native Buildpacks etc.

### Background

There is a current feature request (#45) that proposes for Coolify to be installed on Kubernetes, this Discussion develops on this exploring the challenges, drawbacks and possible solutions. The changes proposed are merely ideas that can be amended during further discussion, if necessary.

### Integration

The first issue to consider (possibly the biggest) is how deeply should Coolify integrate with Kubernetes, there is huge scope for Coolify to play here, whether we simply allow for deployments to be posted to Pods or have Coolify depend on Kubernetes completely. Without a doubt, Kubernetes provides alot of functionality out of the box for Coolify to build on top of. My proposal advocates for tight integration.

One of the primary reasons for this discussion is to consider scale, while most Coolify instances only have a few projects that needs to be built and deployed on a single machine, this is not the case in the enterprise world, where thousands, if not, millions of projects need to be maintained, and in most cases, several versions of it (think production and staging versions). This is where Kubernetes would play its role, Kubernetes has been battle tested for this kind of scale, but is also capable of working on single node installations allowing for a variety of use cases. However Kubernetes is not a fix it all solution, it has its own limitations (further explained here: https://kubernetes.io/docs/setup/best-practices/cluster-large/), for truly large scale clusters, service meshes like Istio would be needed that allow for multi cluster deployments, allowing for endless amounts of scalability. Coolify needs to take this into consideration, as such the Kubernetes installation should integrate with Istio or similar alternatives, to tackle this problem.

Building, this can also be offloaded to Kubernetes with the likes of Kpack (https://github.com/pivotal/kpack/) - again this is largely down to how deeply Coolify would be integrated with Kubernetes, by using a service such as Kpack, it handles the building and tracking for changes, reducing the complexity of Coolify (v2), and allows automatically for building a myriad of types of applications (everything from PHP, Rust, Elm, Node - you name it), and the best part is you can build your own quite easily if a builder does not already exist. For anyone confused on this matter, please research further about "CloudNative Buildpacks" or read up the docs on Kpack.

#### Features

**Cron jobs (#75)**

Kubernetes already allows for Cron jobs to be deployed, this means Coolify would not have to develop any particular system to allow for these, rather build on the Kubernetes API to allow for these to be deployed.

**Multi Host OS (#76)**

Currently Coolify runs on Ubuntu as it needs to execute a multitude of scripts to fetch, build and deploy a project, with Kubernetes this can be offloaded to the K8s control plane. This will reduce the complexity of Coolify as a project, allowing for the team to work on more essential issues not related to scaling or platform.

**Multi tenancy (#77)**

With the usage of Namespaces within Kubernetes, multi tenancy can be achieved, allowing for multiple teams to have different resources, in combination with Users and Roles in Kubernetes, one can achieve true multi tenancy with minimal implementation.

#### Installation

Installation can easily be provided with an Helm chart, and my personal recommendation for the project.

Overall, the idea is to reduce the technical implementation in Coolify while relying on a collection of enterprise grade projects to allow for Coolify to delve into the business logic and allowing for the "user" - Business or Individual running the Coolify instance all the freedom of choice in their tooling.

If you have any questions or concerns, to discuss further on this topic.

**Comment by andrasbacsai** [comment]
Thank you for the detailed explanation! I'm still exploring all the possibilities for the Kubernetes integrations. Your detailed view on the topic is helping a lot! 🙂

I'll add more details to this discussion later on!

**Reply by badlanguage** [reply]
When you get a chance, I would love to hear your take on this, and the approach you decide on for Kubernetes integration

**Reply by adiologydev** [reply]
Any thoughts on this? Just asking because it's been a while :)

**Reply by SamAmann** [reply]
Joigning the interested people in that discussion (even if, indeed, it's been a while :) )

**Comment by badlanguage** [comment]
I would like to bring attention to [Rancher Rio](https://github.com/rancher/rio) since we are discussing about deep integrations with Kubernetes, this is a project Coolify can take some inspiration from and develop a more complete product on, additionally another feature you can add is @andrasbacsai Docker registry based deployments from (custom) registries.

**Comment by tubbo** [comment]
I know this is a pretty old thread, but enabling Coolify integration within Kubernetes is definitely doable right now if you just simply took the `docker-compose.prod.yml` and translated it into a Helm chart or something. That's an overly simplistic approach, but after quickly reading the compose file it doesn't seem like Coolify needs anything "special" from Docker that would prevent it from running on Kubernetes right now, it just needs a chart to make things easier. Technically, it should be possible to use the [bjw-s/app-template](https://github.com/bjw-s/helm-charts/tree/main/charts/other/app-template) chart to run this right now in a self-hosted Kubernetes cluster, right?

**Reply by compgeniuses** [reply]
i think the idea and discussion was mainly in usig coolify to deploy Kubernetes applications, not running coolify in kubernetes
